### PR TITLE
Add functions definitions to the `providers schema -json` output

### PR DESCRIPTION
**Pull request by jbardin**
Add function definitions to the json schema CLI output.  While the protocol names may not exactly match the json keys, this matches the existing `metadata functions` output for easier integration with tooling already using that interface. We use the `jsonfunction` package from the `metadata` command to serialize the new provider functions, and ensure they stay synchronized. This makes the integration a little clumsy, and it may be a good idea to combine the cli json definitions into a single package, but the separate serialization works well enough for now.

An example of the `"functions"` portion of the schema would look like:

```
      "functions": {
        "format_version": "1.0",
        "function_signatures": {
          "noop": {
            "description": "noop takes any single argument and returns the same value",
            "return_type": "dynamic",
            "parameters": [
              {
                "name": "noop",
                "description": "any value",
                "is_nullable": true,
                "type": "dynamic"
              }
            ]
          }
        }
      }
    },
```

The provider functions are not being added to the `metadata functions` command. The function names in the configuration are dependent on the local provider name in the module, and therefor are context dependent. The `metadata` command is also not guaranteed to have an initialized working directory and providers may not be available at all.

### CODE REVIEW

**Comment by jbardin** [comment]
> > While the protocol names may not exactly match the json keys
> 
> Can you expand on this? Is it just the whitespaces around `::` that you're referring to here or is there some other reason for the mismatch?
> 

Simply the field names decided on for the protocol don't align with the existing keys used in the json function definitions, e.g. `is_nullable` != "AllowNullValues"

> > The function names in the configuration are dependent on the local provider name in the module, and therefor are context dependent.
> 
> Is there any value in embedding the local provider name in the output? i.e. are there any consumers who would benefit from it today?

The local provider name could be any valid identifier as defined within any module, and there may be many for any particular provider. Since that is dependent on the configuration, and schema is _independent_ of the configuration, it feels more clear to simply state the schema and leave the scoping to the configuration.

> So far the connection between configuration and `terraform providers schema -json` was only indirect, an implementation detail of how providers get installed. AFAICT In theory we [could have](https://github.com/hashicorp/terraform/issues/24261) a future `providers schema -json` which wholly relies on provider binaries and leaves the complexities of discovering and installing and choosing providers to `init`. This moves us further away from ever achieving that.
> 
> TL;DR Can we just use the last segment such as `count_e` from `provider::aws::count_e` in the JSON output?
> 

Sorry, the json output is only the last segment which is the function name within the provider's schema, because the scoping of the function is not related to the schema. I'll add an example here shortly

**Comment by bflad** [comment]
Other than synchronizing the machine-readable outputs to make it somewhat easier for integrations, is there a better reason to synchronize the two? The provider-defined functions protocol was specifically designed for the additional needs of provider implementation details and documentation, I'm worried that the drift between the naming of things between the protocol and machine-readable output will ultimately be more confusing if/when we change it further. Being able to show fields like `allow_unknown_values` could improve documentation, for example, by having the documentation generation aware of that behavior. Since our protocol definitions tend to remain fairly static, I think setting us up for less potential pain/confusion in the future is a beneficial move, over any potential ease benefits of integrations that need to reference both outputs.

**Comment by radeksimko** [comment]
@bflad I'm not opposed to the provider definitions containing new fields, as long as the existing fields with the same meaning are called the same. That was my main concern in the original RFC that the two formats drift in naming.

That said I also wonder whether these additional fields could be relevant for core functions too, or why don't we expose it today? Is that something that you think would make documentation of Core functions better?

Of course the function docs today are still hand-maintained but I can imagine a future where they become generated too, so this would be IMO relevant question to ask.

**Comment by jbardin** [comment]
If the names are diverging, but must be consumed via the same interface, then _someone_ must translate the divergent structures. I position here was that since core is the point at which the names diverge, it could be responsible for that translation as well. Having a single structure to deal with all function output ensures they are always aligned with future changes as well.

Both core and provider functions follow the same cty calling convention and have the same options. The reason that `allow_unknown_values` was not included originally, is that it is only an implementation detail and makes no difference to the end user which way it is set (it's always "allowed" in the config, it's just whether the call is made to the provider or resolved statically)

**Review by radeksimko — APPROVED** [review]
I don't have a provider handy with which I could test this but I'm choosing to trust the example output you attached reflects the implementation. 😄 i.e. only the function names without the `::` separators and other segments are included in the output.

In which case LGTM, I have no further objections here.

**Review by bflad — APPROVED** [review]
I trust both your opinions here; my only request then would be to ensure https://developer.hashicorp.com/terraform/cli/commands/providers/schema gets updated as well. 👍

**Comment by bflad** [comment]
By the way, the function definition summary field is probably a better example. The existing `jsonfunction` handling doesn't handle that information.

**Comment by jbardin** [comment]
Thanks @bflad, I'll follow up with as addition covering `summary`. That one is new to the protocol.
